### ISSUE TITLE
Update transaction retrieval and confirmation waiting

### ISSUE THREAD

**Issue by helloibis** [ApeWorX/ape#937]
### Overview

To facilitate #251 we need to allow certain functions to optionally raise exceptions for failed transactions _that have been committed_ (i.e. passed preparation, gas estimation, etc. but were actually sent and committed on chain).

### Specification

`ReceiptAPI.await_confirmation()` needs to be able to `raise_for_status()` and have the exception bubble up. This is so the `RevertsContextManager` can catch an exception and handle the transaction failure, because otherwise `raise_for_status()` (the last "piece in the chain" of a transaction where a Python exception can be raised) does not raise an exception, which means the context manager fails.

`ProviderAPI.get_transaction()` then should have the option for the user to skip awaiting receipt confirmation. This way on the first time around (sending the transaction) we do end up awaiting confirmation and raising an exception for `RevertsContextManager` to catch... but then inside the context manager we can call `get_transaction(..., await_confirmations=False)` to get the receipt

The flag should also handle gas estimation, so that you can bypass gas estimation failures and receipt `raise_for_status()` with `raise_on_fail=False`. This makes it easier to bypass estimation over using `gas_limit=...`

### Examples

```python
# Default is that the txn should fail and this should raise (in this case due to gas estimation)
vyper_contract_instance.setNumber(5, sender=owner)

# You can still specify a gas limit to get past that behavior
# With the new flag, it should default to True and a call like this should still raise, but the transaction committed
vyper_contract_instance.setNumber(5, sender=owner, gas_limit=100000)

# raise_on_fail=False should cause the txn not to fail on gas or confirmation
receipt = vyper_contract_instance.setNumber(5, sender=owner, raise_on_fail=False)

# Acts as a middleware that forces a transaction through
# Gas estimation should not fail here but the failed receipt should raise for exception
with ape.reverts():
    vyper_contract_instance.setNumber(5, sender=owner)
```

### Dependencies

https://github.com/ApeWorX/ape/issues/985

**Comment by helloibis** [comment]
Just updated the ticket to include the requirements discussed with Bryant today

**Comment by helloibis** [comment]
Can be completed once https://github.com/ApeWorX/ape/issues/985 is finished

**Comment by helloibis** [comment]
Merged this feature at https://github.com/ApeWorX/ape/pull/957, but it'll hold in a feature branch until we get a release ready for version 0.6.0

**Comment by helloibis** [comment]
This is an old issue, we're no longer taking this approach (and the feature is already done)
